relayer/src/worker/packet.rs

```diff
@@ -144,94 +108,98 @@ pub fn spawn_packet_cmd_worker<ChainA: ChainHandle, ChainB: ChainHandle>(
+/// Whether or not to clear pending packets at this `step` for the given height.
+/// Packets are cleared if `clear_interval` is not `0` and if we have reached the interval.
+fn should_clear_packets(clear_interval: u64, height: Height) -> bool {
+    clear_interval != 0 && height.revision_height % clear_interval == 0
+}
 
-        return RetryResult::Retry(index);
-    }
+fn handle_update_schedule<ChainA: ChainHandle, ChainB: ChainHandle>(
+    link: &mut Link<ChainA, ChainB>,
+    clear_interval: u64,
+    path: &Packet,
+    batch: EventBatch,
+) -> Result<(), TaskError<RunError>> {
+    link.a_to_b
+        .update_schedule(batch)
+        .map_err(handle_link_error_in_task)?;
 
-    // The calls to refresh_schedule and execute_schedule depends on
-    // earlier calls to update_schedule and schedule_packet_clearing.
-    // Hence they must be retried in the same function body so that
-    // the same WorkerCmd is used for retrying the whole execution.
-    //
-    // The worker for spawn_packet_worker is still needed to handle
-    // the case when no PacketCmd arriving, so that it can still
-    // do the refresh and execute schedule.
-    // This follows the original logic here:
-    // https://github.com/informalsystems/ibc-rs/blob/e7a6403888f48754ddb80e35ebe2281fb7c51c04/relayer/src/worker/packet.rs#L127-L133
-
-    let schedule_result = link
-        .a_to_b
+    handle_execute_schedule(link, path, Resubmit::from_clear_interval(clear_interval))
+}
+
+fn handle_clear_packet<ChainA: ChainHandle, ChainB: ChainHandle>(
+    link: &mut Link<ChainA, ChainB>,
+    clear_interval: u64,
+    path: &Packet,
+    height: Option<Height>,
+) -> Result<(), TaskError<RunError>> {
+    link.a_to_b
+        .schedule_packet_clearing(height)
+        .map_err(handle_link_error_in_task)?;
+
+    handle_execute_schedule(link, path, Resubmit::from_clear_interval(clear_interval))
+}
+
+fn handle_execute_schedule<ChainA: ChainHandle, ChainB: ChainHandle>(
+    link: &mut Link<ChainA, ChainB>,
+    _path: &Packet,
+    resubmit: Resubmit,
+) -> Result<(), TaskError<RunError>> {
+    link.a_to_b
         .refresh_schedule()
-        .and_then(|_| link.a_to_b.execute_schedule());
-
-    if let Err(e) = schedule_result {
-        if e.is_expired_or_frozen_error() {
-            error!("aborting due to expired or frozen client");
-            return RetryResult::Err(index);
-        } else {
-            error!(
-                retry_index = %index,
-                "will retry: schedule execution encountered error: {}",
-                e,
-            );
-            return RetryResult::Retry(index);
-        }
-    }
+        .map_err(handle_link_error_in_task)?;
+
+    link.a_to_b
+        .execute_schedule()
+        .map_err(handle_link_error_in_task)
+        .map_err(|e| {
+            if let TaskError::Ignore(e) = &e {
+                error!("will retry: schedule execution encountered error: {}", e,);
+            }
+            e
+        })?;
```

Comment:
I'm not really a fan of the back-to-back `map_err` calls; they're a bit confusing to trace through. Could we do something like this instead?

```suggestion
link.a_to_b.execute_schedule().map_err(|e| {
  let task_error = handle_link_error_in_task(e);
  
  if let TaskError::Ignore(e) = &task_error {
    error!("will retry: schedule execution encountered error: {}", e,);
  }
  
  task_error
})?;
```